File: old/TNode.java
import java.util.ArrayList;

public class TNode<E> extends Node<E> {
	protected TNode<E> parent;
	protected ArrayList<TNode<E>> children;
	
	public TNode() {
		this.element = null;
		this.parent = null;
		this.children = new ArrayList<TNode<E>>();
	}

	public TNode(E element) {
		this.element = element;
		this.parent = null;
		this.children = new ArrayList<TNode<E>>();
	}

	public TNode(TNode<E> parent) {
		this.parent = parent;
		this.element = null;
		this.children = new ArrayList<TNode<E>>();
	}

	public TNode(E element, TNode<E> parent) {
		this.element = element;
		this.parent = parent;
		this.children = new ArrayList<TNode<E>>();
	}

	public TNode<E> getParent() {
		return this.parent;
	}

	public void setParent(TNode<E> p) {
		this.parent = p;
	}

	public ArrayList<TNode<E>> getChildren() {
		return this.children;
	}

	public void setChildren(ArrayList<TNode<E>> children) {
		this.children = children;
	}

	public boolean isExternal() {
		return children.isEmpty();
	}

	public boolean isInternal() {
		return(!isExternal());
	}
}

File: c++/oldy/older/older/TNode.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include "Node.h"
#include "TNode.h"

int main() {
	TNode<string> n;
	TNode<string> p;

	n.setElement("David III");
	p.setElement("Martina Teran");

	cout << "P is " << p.getElement() << endl;
	cout << "N is " << n.getElement() << endl;
	
	p.addChild(n);
	n.setParent(p);
	
	if(n.isExternal()) {
		cout << "N is external!\n";
	}
	
	if(p.isExternal()) {
		cout << "P is external!\n";
	}
	
	if(p.isInternal()) {
		cout << "P is internal!\n";
	}
	
	TNode<string> &parent = n.getParent();
	cout << "Parent is " << parent.getElement() << endl;
	
	p.setIndex(0);
	n.setIndex(1);
	
	cout << "P index is " << p.getIndex() << endl;
	cout << "N index is " << n.getIndex() << endl;
	
	TNode<string> &pchild = p.getChild(0);
	
	cout << "Pchild is " << pchild.getElement() << endl;
	
	TNode<string> &nparent = n.getParent();
	
	cout << "nparent is " << nparent.getElement() << endl;
}
